driver-async/src/main/com/mongodb/async/client/FetchAllObserver.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.async.client;
+
+class FetchAllObserver<TResult> implements Observer<TResult> {
+
+    private final Observer<TResult> delegate;
+    private final long batchSize;
+    private long counter = 0;
+    private Subscription subscription;
+
+    public FetchAllObserver(final Observer<TResult> delegate, final Integer batchSize) {
```

Comment:
The reason I included batchSize is a) because if a user wants to granularly control those batches then they can and b) this is the current behaviour for `into` and `foreach`.

As `request(long n)` is directly coupled to `batchSize` we can't request `Long.MAX_VALUE` as that would set the `batchSize` to `Integer.MAX_VALUE` and the user may have reasons not to want to do that - eg start processing results quickly rather and do multiple io trips.

In reality the default `batchSize` if not set will be `Integer.MAX_VALUE` so it won't make much difference but the user can override if desired.